File: gui/lib/models/motion.dart
import 'dart:convert';

String motionsToJson(List<Motion> motions) =>
    jsonEncode(motions.map((Motion motion) => motion.toJson()).toList());

class Motion {
  int leg = 0;
  int startTime = 0;
  int duration = 0;

  double q = 0;
  double dq = 0;
  double tau = 0;
  double kp = 0;
  double kd = 0;

  Motion({
    this.leg = 0,
    this.startTime = 0,
    this.duration = 0,
    this.q = 0,
    this.dq = 0,
    this.tau = 0,
    this.kp = 0,
    this.kd = 0,
  });

  factory Motion.fromJson(Map<String, dynamic> json) => Motion(
        leg: json["leg"],
        startTime: json["start_time"],
        duration: json["duration"],
        q: json["q"].toDouble(),
        dq: json["dq"].toDouble(),
        tau: json["tau"].toDouble(),
        kp: json["kd"].toDouble(),
        kd: json["kp"].toDouble(),
      );

  Map<String, dynamic> toJson() => {
        "leg": leg,
        "start_time": startTime,
        "duration": duration,
        "q": q,
        "dq": dq,
        "tau": tau,
        "kp": kp,
        "kd": kd,
      };
}

File: dog/src/models/motion.cc
#include <vector>

#include <drogon/drogon.h>
#include <server/models/motion.h>
#include <iostream>

namespace drogon
{
    template<>
    std::vector<Server::Motion> fromRequest(const drogon::HttpRequest &req)
    { 
        std::vector<Server::Motion> motions = {};
        auto json = req.getJsonObject();

        for (auto json_part : (*json))
        {
            Server::Motion motion;

            motion.leg = json_part["leg"].asUInt();
            motion.start_time = json_part["start_time"].asUInt() + 50;
            motion.duration = json_part["duration"].asUInt();
            motion.sleep_time = json_part["sleep_time"].asUInt();
            
            motion.q = json_part["q"].asFloat();
            motion.dq = json_part["dq"].asFloat();
            motion.tau = json_part["tau"].asFloat();
            motion.Kp = json_part["kp"].asFloat();
            motion.Kd = json_part["kd"].asFloat();

            motions.push_back(motion);
        }
        // std::cout << (*json) << std::endl;

        return motions;
    }
}
